python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py

```diff
@@ -0,0 +1,1124 @@
+
+        self.l2ScalarProductMatrix = None
+        self.h1ScalarProductMatrix = None
+        self.l2ProductBasis = [] # list containing the vector given the column of (l2ScalarProductMatrix @ reeducedBasis) 
+        self.reducedBasis = None # list containing the vector of reduced basis function
+        self.N = 0 # number of modes
+
+        if self.doRectification or initCoarse:
+            super().initCoarseToolbox()
+        if self.worldcomm.isMasterRank():
+            print(f"[NIRB] Initialization done")
+
+
+
+    def BiOrthonormalization(self):
+        """Bi-orthonormalization of reduced basis
+        """
+        K = np.zeros((self.N,self.N))
+        M = K.copy()
+
+        for i in range(self.N):
+            for j in range(self.N):
+                K[i,j] = self.h1ScalarProductMatrix.energy(self.reducedBasis[i],self.reducedBasis[j])
+                M[i,j] = self.l2ScalarProductMatrix.energy(self.reducedBasis[i],self.reducedBasis[j])
+
+        from scipy import linalg
+
+        eval,evec=linalg.eigh(a=K, b=M, overwrite_a=True, overwrite_b=True) #eigenvalues
+        eigenValues = eval.real
+        eigenVectors = evec 
+        idx = eigenValues.argsort()[::-1]
+        eigenValues = eigenValues[idx]
+        eigenVectors = evec[:, idx]
+
+        for i in range(self.N):
+            eigenVectors[:,i] /= np.sqrt(np.abs(eigenValues[i]))
+
+        oldbasis = self.reducedBasis.copy()
+        self.reducedBasis = []
+
+        for i in range(self.N):
+            vec = self.Xh.element()
+            vec.setZero()
+            for j in range(self.N):
+                vec = vec + float(eigenVectors[j,i])*oldbasis[j]
+                
+            self.reducedBasis.append(vec)
+
+
+    def initProblem(self, numberOfInitSnapshots, Xi_train=None, samplingMode="log-random", computeCoarse=False):
+        """Initialize the problem
+
+        Args:
+        -----
+            numberOfInitSnapshots (int): number of snapshots to use for the initialization
+            Xi_train (list of ParameteœrSpaceElement, optional): Train set for algorithm. If None is given, a set of size Ntrain is generated. Defaults to None.
+            samplingMode (str, optional): sampling mode in the parameter space.(random, log-random, log-equidistribute, equidistribute) Defaults to "log-random".
+            computeCoarse (bool, optional): compute snapshots for coarse toolbox, used for rectification. Defaults to False.
+        
+        Returns:
+        --------
+            Xi_train (list of ParameterSpaceElement): parameters used for do build the basis
+        """
+        if self.doRectification:
+            computeCoarse=True
+
+        self.fineSnapShotList = []
+        self.coarseSnapShotList = []
+
+        if Xi_train is None:
+            s = self.Dmu.sampling()
+            s.sampling(numberOfInitSnapshots, samplingMode)
+            vector_mu = s.getVector()
+        else:
+            vector_mu = choices(Xi_train, k=numberOfInitSnapshots)
```

Comment:
there is an issue in the function choices. It doesn't return what we expect. 
I give a list of 10 params and ask him to chose 100 params. It didn't crash !!

Reply:
It seems that `random.choices` runs the draw with replacement. To avoid it you can use the numpy function :
```python
np.random.choice(list, size=size, replace=False)
```